jenkins-design-language/src/js/components/pipeline/PipelineGraphLayout.jsx

```diff
@@ -0,0 +1,230 @@
+    return {
+        nodeColumns: allNodeColumns,
+        connections,
+        bigLabels,
+        smallLabels,
+        measuredWidth,
+        measuredHeight,
+    };
+}
+
+/**
+ * Generate an array of columns, based on the top-level stages
+ */
+function createNodeColumns(topLevelStages: Array<StageInfo> = []): Array<NodeColumn> {
+
+    const nodeColumns = [];
+
+    for (const topStage of topLevelStages) {
+        // If stage has children, we don't draw a node for it, just its children
+        const stagesForColumn =
+            topStage.children && topStage.children.length ? topStage.children : [topStage];
+
+        const column = {
+            topStage,
+            nodes: [],
+        };
+
+        column.nodes = stagesForColumn
+            .filter(nodeStage => !!nodeStage)
+            .map(nodeStage => ({
+                x: 0, // Layout is done later
+                y: 0,
+                name: nodeStage.name,
+                id: nodeStage.id,
+                stage: nodeStage,
+                isPlaceholder: false,
+                key: 'n_' + nodeStage.id,
+            }));
+
+        nodeColumns.push(column);
+    }
+
+    return nodeColumns;
+}
+
+
+/**
+ * Walks the columns of nodes giving them x and y positions. Mutates the node objects in place for now.
+ */
+function positionNodes(nodeColumns: Array<NodeColumn>, { nodeSpacingH, nodeSpacingV, ypStart }) {
+
+    let xp = nodeSpacingH / 2;
+    let previousTopNode = null;
+
+    for (const column of nodeColumns) {
+        const topNode = column.nodes[0];
+
+        let yp = ypStart; // Reset Y to top for each column
+
+        if (previousTopNode) {
+            // Advance X position
+            if (previousTopNode.isPlaceholder || topNode.isPlaceholder) {
+                // Don't space placeholder nodes (start/end) as wide as normal.
+                xp += Math.floor(nodeSpacingH * 0.7);
+            } else {
+                xp += nodeSpacingH;
+            }
+        }
+
+        for (const node of column.nodes) {
+            node.x = xp;
+            node.y = yp;
+
+            yp += nodeSpacingV;
+        }
+
+        previousTopNode = topNode;
+    }
+}
+
+
+/**
+ * Generate label descriptions for big labels at the top of each column
+ */
+function createBigLabels(columns: Array<NodeColumn>) {
+
+    const labels = [];
+
+    for (const column of columns) {
+
+        const node = column.nodes[0];
+        const stage = column.topStage;
+        const text = stage ? stage.name : node.name;
+        const key = 'l_b_' + node.key;
+
+        labels.push({
+            x: node.x,
+            y: node.y,
+            node,
+            stage,
+            text,
+            key,
+        });
+    }
+
+    return labels;
+}
+
+/**
+ * Generate label descriptions for small labels under the nodes
+ */
+function createSmallLabels(columns: Array<NodeColumn>) {
+
+    const labels = [];
+
+    for (const column of columns) {
+        if (column.nodes.length === 1) {
```

Comment:
Should we null check `column.nodes` first?

Reply:
That's `flow`'s job :) All the columns are generated within this module, they're not inputs so if it's bolloxed I want errors.